File: game-development/pathfinding.py
"""
Pathfinding - Python

Finding paths in game worlds:
- A* Algorithm (game-specific)
- Navigation mesh
- Grid-based pathfinding
"""

from typing import List, Tuple, Optional, Dict, Set
import heapq


# Grid-based A* Pathfinding
class GridPathfinder:
    def __init__(self, grid: List[List[int]]):
        self.grid = grid  # 2D array: 0 = walkable, 1 = obstacle
        self.width = len(grid[0])
        self.height = len(grid)
    
    def heuristic(self, a: Tuple[int, int], b: Tuple[int, int]) -> float:
        # Manhattan distance
        return abs(a[0] - b[0]) + abs(a[1] - b[1])
    
    def get_neighbors(self, node: Tuple[int, int]) -> List[Tuple[int, int]]:
        neighbors = []
        directions = [
            (0, -1),  # Up
            (1, 0),   # Right
            (0, 1),   # Down
            (-1, 0)   # Left
        ]
        
        for dx, dy in directions:
            x = node[0] + dx
            y = node[1] + dy
            
            if 0 <= x < self.width and 0 <= y < self.height:
                if self.grid[y][x] == 0:  # Walkable
                    neighbors.append((x, y))
        
        return neighbors
    
    def find_path(self, start: Tuple[int, int], goal: Tuple[int, int]) -> Optional[List[Tuple[int, int]]]:
        open_set = [(0, start)]
        came_from: Dict[Tuple[int, int], Optional[Tuple[int, int]]] = {}
        g_score: Dict[Tuple[int, int], float] = {start: 0}
        f_score: Dict[Tuple[int, int], float] = {start: self.heuristic(start, goal)}
        visited: Set[Tuple[int, int]] = set()
        
        while open_set:
            current_f, current = heapq.heappop(open_set)
            
            if current in visited:
                continue
            
            visited.add(current)
            
            if current == goal:
                # Reconstruct path
                path = []
                node = current
                while node is not None:
                    path.append(node)
                    node = came_from.get(node)
                return path[::-1]
            
            for neighbor in self.get_neighbors(current):
                if neighbor in visited:
                    continue
                
                tentative_g = g_score[current] + 1
                
                if neighbor not in g_score or tentative_g < g_score[neighbor]:
                    came_from[neighbor] = current
                    g_score[neighbor] = tentative_g
                    f_score[neighbor] = tentative_g + self.heuristic(neighbor, goal)
                    heapq.heappush(open_set, (f_score[neighbor], neighbor))
        
        return None  # No path found


# Navigation Mesh (simplified)
class NavMesh:
    def __init__(self):
        self.nodes: List[Dict] = []
        self.edges: List[Dict] = []
    
    def add_node(self, x: float, y: float) -> Dict:
        node = {'id': len(self.nodes), 'x': x, 'y': y}
        self.nodes.append(node)
        return node
    
    def add_edge(self, node1: Dict, node2: Dict):
        self.edges.append({'from': node1['id'], 'to': node2['id']})
    
    def find_path(self, start_node: Dict, end_node: Dict) -> Optional[List[Dict]]:
        # Simplified pathfinding on nav mesh
        path = [start_node]
        
        # In real implementation, use A* or Dijkstra on the graph
        if self.has_connection(start_node['id'], end_node['id']):
            path.append(end_node)
            return path
        
        return None
    
    def has_connection(self, node1_id: int, node2_id: int) -> bool:
        return any(
            (e['from'] == node1_id and e['to'] == node2_id) or
            (e['from'] == node2_id and e['to'] == node1_id)
            for e in self.edges
        )


# Example usage
if __name__ == "__main__":
    print("=== Pathfinding ===\n")
    
    # Grid-based pathfinding
    grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 0, 0, 0]
    ]
    
    pathfinder = GridPathfinder(grid)
    start = (0, 0)
    goal = (4, 4)
    
    print("Grid-based A* Pathfinding:")
    print("Grid (0=walkable, 1=obstacle):")
    for y, row in enumerate(grid):
        display_row = []
        for x, cell in enumerate(row):
            if (x, y) == start:
                display_row.append('S')
            elif (x, y) == goal:
                display_row.append('G')
            else:
                display_row.append(str(cell))
        print(f"  {' '.join(display_row)}")
    
    path = pathfinder.find_path(start, goal)
    if path:
        print("\nPath found:")
        for i, node in enumerate(path):
            print(f"  Step {i}: {node}")
    else:
        print("\nNo path found")
    
    # Navigation Mesh
    print("\nNavigation Mesh:")
    nav_mesh = NavMesh()
    node1 = nav_mesh.add_node(0, 0)
    node2 = nav_mesh.add_node(2, 2)
    node3 = nav_mesh.add_node(4, 4)
    nav_mesh.add_edge(node1, node2)
    nav_mesh.add_edge(node2, node3)
    
    nav_path = nav_mesh.find_path(node1, node3)
    if nav_path:
        path_str = " -> ".join(f"({n['x']}, {n['y']})" for n in nav_path)
        print(f"Nav mesh path: {path_str}")

File: game-development/pathfinding.js
/**
 * Pathfinding - JavaScript
 * 
 * Finding paths in game worlds:
 * - A* Algorithm (already in graph-algorithms, but game-specific)
 * - Navigation mesh
 * - Grid-based pathfinding
 */

// Grid-based A* Pathfinding
class GridPathfinder {
    constructor(grid) {
        this.grid = grid; // 2D array: 0 = walkable, 1 = obstacle
        this.width = grid[0].length;
        this.height = grid.length;
    }
    
    // Heuristic function (Manhattan distance)
    heuristic(a, b) {
        return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
    }
    
    // Get neighbors
    getNeighbors(node) {
        const neighbors = [];
        const directions = [
            { x: 0, y: -1 }, // Up
            { x: 1, y: 0 },  // Right
            { x: 0, y: 1 },  // Down
            { x: -1, y: 0 }  // Left
        ];
        
        for (const dir of directions) {
            const x = node.x + dir.x;
            const y = node.y + dir.y;
            
            if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
                if (this.grid[y][x] === 0) { // Walkable
                    neighbors.push({ x, y });
                }
            }
        }
        
        return neighbors;
    }
    
    // A* pathfinding
    findPath(start, goal) {
        const openSet = [start];
        const cameFrom = new Map();
        const gScore = new Map();
        const fScore = new Map();
        
        const startKey = `${start.x},${start.y}`;
        const goalKey = `${goal.x},${goal.y}`;
        
        gScore.set(startKey, 0);
        fScore.set(startKey, this.heuristic(start, goal));
        
        while (openSet.length > 0) {
            // Find node with lowest fScore
            let current = openSet.reduce((min, node) => {
                const nodeKey = `${node.x},${node.y}`;
                const minKey = `${min.x},${min.y}`;
                return fScore.get(nodeKey) < fScore.get(minKey) ? node : min;
            });
            
            const currentKey = `${current.x},${current.y}`;
            
            if (currentKey === goalKey) {
                // Reconstruct path
                const path = [];
                let node = current;
                while (node) {
                    path.unshift(node);
                    const nodeKey = `${node.x},${node.y}`;
                    node = cameFrom.get(nodeKey);
                }
                return path;
            }
            
            openSet.splice(openSet.indexOf(current), 1);
            
            for (const neighbor of this.getNeighbors(current)) {
                const neighborKey = `${neighbor.x},${neighbor.y}`;
                const tentativeGScore = gScore.get(currentKey) + 1;
                
                if (!gScore.has(neighborKey) || tentativeGScore < gScore.get(neighborKey)) {
                    cameFrom.set(neighborKey, current);
                    gScore.set(neighborKey, tentativeGScore);
                    fScore.set(neighborKey, tentativeGScore + this.heuristic(neighbor, goal));
                    
                    if (!openSet.find(n => `${n.x},${n.y}` === neighborKey)) {
                        openSet.push(neighbor);
                    }
                }
            }
        }
        
        return null; // No path found
    }
}

// Navigation Mesh (simplified)
class NavMesh {
    constructor() {
        this.nodes = [];
        this.edges = [];
    }
    
    addNode(x, y) {
        const node = { id: this.nodes.length, x, y };
        this.nodes.push(node);
        return node;
    }
    
    addEdge(node1, node2) {
        this.edges.push({ from: node1.id, to: node2.id });
    }
    
    findPath(startNode, endNode) {
        // Simplified pathfinding on nav mesh
        const path = [startNode];
        const visited = new Set([startNode.id]);
        
        const current = startNode;
        // In real implementation, use A* or Dijkstra on the graph
        if (this.hasConnection(current.id, endNode.id)) {
            path.push(endNode);
            return path;
        }
        
        return null;
    }
    
    hasConnection(node1Id, node2Id) {
        return this.edges.some(e => 
            (e.from === node1Id && e.to === node2Id) ||
            (e.from === node2Id && e.to === node1Id)
        );
    }
}

// Example usage
console.log('=== Pathfinding ===\n');

// Grid-based pathfinding
const grid = [
    [0, 0, 0, 0, 0],
    [0, 1, 1, 1, 0],
    [0, 0, 0, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 0, 0, 0]
];

const pathfinder = new GridPathfinder(grid);
const start = { x: 0, y: 0 };
const goal = { x: 4, y: 4 };

console.log('Grid-based A* Pathfinding:');
console.log('Grid (0=walkable, 1=obstacle):');
grid.forEach((row, y) => {
    console.log(`  ${row.map((cell, x) => {
        if (x === start.x && y === start.y) return 'S';
        if (x === goal.x && y === goal.y) return 'G';
        return cell;
    }).join(' ')}`);
});

const path = pathfinder.findPath(start, goal);
if (path) {
    console.log('\nPath found:');
    path.forEach((node, i) => {
        console.log(`  Step ${i}: (${node.x}, ${node.y})`);
    });
} else {
    console.log('\nNo path found');
}

// Navigation Mesh
console.log('\nNavigation Mesh:');
const navMesh = new NavMesh();
const node1 = navMesh.addNode(0, 0);
const node2 = navMesh.addNode(2, 2);
const node3 = navMesh.addNode(4, 4);
navMesh.addEdge(node1, node2);
navMesh.addEdge(node2, node3);

const navPath = navMesh.findPath(node1, node3);
if (navPath) {
    console.log('Nav mesh path:', navPath.map(n => `(${n.x}, ${n.y})`).join(' -> '));
}

module.exports = { GridPathfinder, NavMesh };

